### ISSUE TITLE
Run out of memory when sending lots of events to http_puma

### ISSUE THREAD

**Issue by jamesweb1** [repeatedly/fluent-plugin-http-puma#1]
I use Apache ab to test the performance of my environment, and I test with following commands:
ab -n 100000 -c 30 -T 'application/json' -S -p data.json [URL]

It works fine when using "-n 10000", but it almost consumes all of my 8GB memory in fluentd server. And I try to increase the number of requests, I find my fluentd restarts as follows message:
2014-07-24 10:18:29 +0000 [error]: fluentd main process died unexpectedly. restarting.

Could anyone tell me why this happens?

**Comment by repeatedly** [comment]
I'm not sure. I tried your ab command but Fluentd works on my VM environment, 1 CPU, 2GB memory.

``` json
{"k1":"1234567gfgfdsadjsalrjeirueore9are9rfua0ojdijf","k2":12345678}
```

``` shell
% ab -n 100000 -c 30 -T 'application/json' -S -p test.json http://127.0.0.1:24555/test.foo
This is ApacheBench, Version 2.3 <$Revision: 655654 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)
Completed 10000 requests
Completed 20000 requests
Completed 30000 requests
Completed 40000 requests
Completed 50000 requests
Completed 60000 requests
Completed 70000 requests
Completed 80000 requests
Completed 90000 requests
Completed 100000 requests
Finished 100000 requests


Server Software:
Server Hostname:        127.0.0.1
Server Port:            24555

Document Path:          /test.foo
Document Length:        2 bytes

Concurrency Level:      30
Time taken for tests:   16.126 seconds
Complete requests:      100000
Failed requests:        0
Write errors:           0
Total transferred:      8500000 bytes
Total POSTed:           21300000
HTML transferred:       200000 bytes
Requests per second:    6201.26 [#/sec] (mean)
Time per request:       4.838 [ms] (mean)
Time per request:       0.161 [ms] (mean, across all concurrent requests)
Transfer rate:          514.75 [Kbytes/sec] received
                        1289.91 kb/s sent
                        1804.66 kb/s total

Connection Times (ms)
              min   avg   max
Connect:        0     0    4
Processing:     1     5   22
Total:          1     5   26

Percentage of the requests served within a certain time (ms)
  50%      5
  66%      5
  75%      5
  80%      5
  90%      5
  95%      6
  98%      6
  99%      7
 100%     26 (longest request)
```

``` shell
% fluentd -c puma.conf
2014-08-04 21:44:22 -0700 [info]: starting fluentd-0.10.52
2014-08-04 21:44:22 -0700 [info]: reading config file path="puma.conf"
2014-08-04 21:44:22 -0700 [info]: gem 'fluent-plugin-flowcounter-simple' version '0.0.3'
2014-08-04 21:44:22 -0700 [info]: gem 'fluent-plugin-http-puma' version '0.0.1'
2014-08-04 21:44:22 -0700 [info]: gem 'fluentd' version '0.10.52'
2014-08-04 21:44:22 -0700 [info]: using configuration file: <ROOT>
  <source>
    type http_puma
    port 24555
  </source>
  <match test.**>
    type flowcounter_simple
    unit second
  </match>
  <match **>
    type null
  </match>
</ROOT>
2014-08-04 21:44:22 -0700 [info]: adding source type="http_puma"
2014-08-04 21:44:22 -0700 [info]: adding match pattern="test.**" type="flowcounter_simple"
2014-08-04 21:44:22 -0700 [info]: adding match pattern="**" type="null"
2014-08-04 21:44:22 -0700 [info]: listening http on 0.0.0.0:24555
2014-08-04 21:44:28 -0700 [info]: plugin:out_flowcounter_simple count:5327      indicator:num   unit:second
2014-08-04 21:44:29 -0700 [info]: plugin:out_flowcounter_simple count:6339      indicator:num   unit:second
2014-08-04 21:44:30 -0700 [info]: plugin:out_flowcounter_simple count:5872      indicator:num   unit:second
2014-08-04 21:44:31 -0700 [info]: plugin:out_flowcounter_simple count:6341      indicator:num   unit:second
2014-08-04 21:44:32 -0700 [info]: plugin:out_flowcounter_simple count:6365      indicator:num   unit:second
2014-08-04 21:44:33 -0700 [info]: plugin:out_flowcounter_simple count:5732      indicator:num   unit:second
2014-08-04 21:44:34 -0700 [info]: plugin:out_flowcounter_simple count:6304      indicator:num   unit:second
2014-08-04 21:44:35 -0700 [info]: plugin:out_flowcounter_simple count:6343      indicator:num   unit:second
2014-08-04 21:44:36 -0700 [info]: plugin:out_flowcounter_simple count:6245      indicator:num   unit:second
2014-08-04 21:44:37 -0700 [info]: plugin:out_flowcounter_simple count:6244      indicator:num   unit:second
2014-08-04 21:44:38 -0700 [info]: plugin:out_flowcounter_simple count:6316      indicator:num   unit:second
2014-08-04 21:44:39 -0700 [info]: plugin:out_flowcounter_simple count:6326      indicator:num   unit:second
2014-08-04 21:44:40 -0700 [info]: plugin:out_flowcounter_simple count:6227      indicator:num   unit:second
2014-08-04 21:44:41 -0700 [info]: plugin:out_flowcounter_simple count:6077      indicator:num   unit:second
2014-08-04 21:44:42 -0700 [info]: plugin:out_flowcounter_simple count:5539      indicator:num   unit:second
2014-08-04 21:44:43 -0700 [info]: plugin:out_flowcounter_simple count:6358      indicator:num   unit:second
2014-08-04 21:44:44 -0700 [info]: plugin:out_flowcounter_simple count:2045      indicator:num   unit:second
```

**Comment by repeatedly** [comment]
Ah wait. Do you use HTTPS?

**Comment by repeatedly** [comment]
Hmm. I tried HTTPS but Fluentd also works on same environment.

``` shell
/usr/lib/go/bin/boom -allow-insecure -n 100000 -c 30 -T 'application/json' -m POST -d '{"k1":"1234567gfgfdsadjsalrjeirueore9are9rfua0ojdijf","k2":12345678}' https://localhost:24555/test.foo
```

I use boom because ab have a problem with HTTPS.

``` shell
% fluentd -c puma.conf
2014-08-04 22:52:08 -0700 [info]: starting fluentd-0.10.52
2014-08-04 22:52:08 -0700 [info]: reading config file path="puma.conf"
2014-08-04 22:52:08 -0700 [info]: gem 'fluent-plugin-flowcounter-simple' version '0.0.3'
2014-08-04 22:52:08 -0700 [info]: gem 'fluent-plugin-http-puma' version '0.0.1'
2014-08-04 22:52:08 -0700 [info]: gem 'fluentd' version '0.10.52'
2014-08-04 22:52:08 -0700 [info]: using configuration file: <ROOT>
  <source>
    type http_puma
    port 24555
    use_ssl true
    ssl_keys ["/home/repeatedly/puma-test/server.key", "/home/repeatedly/puma-test/server.cert"]
  </source>
  <match test.**>
    type flowcounter_simple
    unit second
  </match>
  <match **>
    type null
  </match>
</ROOT>
2014-08-04 22:52:08 -0700 [info]: adding source type="http_puma"
2014-08-04 22:52:08 -0700 [info]: adding match pattern="test.**" type="flowcounter_simple"
2014-08-04 22:52:08 -0700 [info]: adding match pattern="**" type="null"
2014-08-04 22:52:08 -0700 [info]: listening https on 0.0.0.0:24555
2014-08-04 22:52:10 -0700 [info]: plugin:out_flowcounter_simple count:75        indicator:num   unit:second
2014-08-04 22:52:11 -0700 [info]: plugin:out_flowcounter_simple count:4262      indicator:num   unit:second
2014-08-04 22:52:12 -0700 [info]: plugin:out_flowcounter_simple count:4348      indicator:num   unit:second
2014-08-04 22:52:13 -0700 [info]: plugin:out_flowcounter_simple count:4028      indicator:num   unit:second
2014-08-04 22:52:14 -0700 [info]: plugin:out_flowcounter_simple count:4193      indicator:num   unit:second
2014-08-04 22:52:15 -0700 [info]: plugin:out_flowcounter_simple count:4455      indicator:num   unit:second
2014-08-04 22:52:16 -0700 [info]: plugin:out_flowcounter_simple count:3788      indicator:num   unit:second
... same logs
```

**Comment by jamesweb1** [comment]
Hi repeatedly,
I also use boom to measure my performance and cannot reproduce my issue as well. My command is boom -n 2000000 -c 30 -T 'application/json' -m POST -d data_60k.json https:[URL]:[port]/count/app

BUT, boom seems to a little strange:
-c 30: ~300 request/sec
-c 300: ~1100 request/sec
As my previous test, it shouldn't so high when we use https/ssl.

However, I can reproduce my issue by Apache Benchmark. I know the issue of ab is the error message "SSL read failed - closing connection", but it seems to not affect the result of benchmark so that I trust that(250 requests/sec). My command is...
ab -n 100000 -c 30  -T 'application/json' -p data_std.json https://[URL]:[port]/count/app
It consumed the memory dramatically as attached picture.
![issue_memory](https://cloud.githubusercontent.com/assets/6868915/3808064/c686010c-1c71-11e4-9a7a-d55d24ff31de.jpg)

And, I also have my own java https client to send messages to fluentd with ssl. It also consumed the memory of fluentd server dramatically so that I post this issue. Do you have any idea on that?

**Comment by repeatedly** [comment]
You can check what object consumes the memory using `in_object_space` plugin.

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_object_space.rb

``` aconf
<source>
  type object_space
  emit_interval 1s # 1 seconds
  tag os.status
</source>

<match os.**>
  type stdout
</match>
```

Could you try this?

I assume when long keepalive or clean up processing from client is bad, puma keeps its connection and related objects. But I'm not sure the cause for now...

**Comment by jamesweb1** [comment]
When I try this plugin, it shows:
2014-08-05 08:44:19 +0000 os.status: {"String":26869,"RubyVM::InstructionSequence":5238,"Array":5048,"Class":997,"Hash":947,"Proc":612,"RubyVM::Env":547,"Regexp":528,"Time":499,"Gem::Version":203,"Gem::Requirement":188,"Thread::Backtrace":172,"Module":167,"IO::EAGAINWaitReadable":167,"Puma::MiniSSL::Engine":143}
//------------------------------------------------------
2014-08-05 08:44:20 +0000 os.status: {"String":28277,"Array":5325,"RubyVM::InstructionSequence":5238,"Hash":1008,"Class":997,"Proc":629,"Time":599,"RubyVM::Env":564,"Regexp":528,"Thread::Backtrace":206,"Gem::Version":203,"IO::EAGAINWaitReadable":201,"Gem::Requirement":188,"Puma::MiniSSL::Engine":184,"Puma::HttpParser":183}
//------------------------------------------------------
2014-08-05 08:44:21 +0000 os.status: {"String":30292,"Array":5833,"RubyVM::InstructionSequence":5238,"Hash":1105,"Class":997,"Time":705,"Proc":672,"RubyVM::Env":607,"Regexp":528,"Thread::Backtrace":240,"Puma::MiniSSL::Engine":237,"Puma::MiniSSL::Socket":236,"Puma::HttpParser":236,"TCPSocket":236,"Puma::Client":236}
//------------------------------------------------------
2014-08-05 08:44:22 +0000 os.status: {"String":32610,"Array":6479,"RubyVM::InstructionSequence":5238,"Hash":1212,"Class":997,"Time":824,"Proc":720,"RubyVM::Env":655,"Regexp":528,"Puma::MiniSSL::Engine":297,"Puma::HttpParser":295,"TCPSocket":295,"Puma::MiniSSL::Socket":295,"Puma::Client":295,"StringIO":272}
//------------------------------------------------------
2014-08-05 08:44:23 +0000 os.status: {"String":34303,"Array":6891,"RubyVM::InstructionSequence":5238,"Hash":1296,"Class":997,"Time":914,"Proc":759,"RubyVM::Env":694,"Regexp":528,"TCPSocket":340,"Puma::MiniSSL::Engine":340,"Puma::MiniSSL::Socket":340,"Puma::HttpParser":340,"Puma::Client":340,"StringIO":316}
//------------------------------------------------------
2014-08-05 08:44:24 +0000 os.status: {"String":24669,"Array":6270,"RubyVM::InstructionSequence":5238,"Class":997,"Hash":784,"Regexp":528,"Proc":393,"RubyVM::Env":328,"Time":241,"Gem::Version":203,"Puma::MiniSSL::Engine":200,"TCPSocket":194,"Puma::HttpParser":194,"Puma::MiniSSL::Socket":194,"Puma::Client":194}
//------------------------------------------------------
Actually, I don't really understand the meaning of these numbers. I just use ab for several seconds and it consumed 2GB memory of fluentd server......

**Comment by repeatedly** [comment]
It seems keepalive problem.
When a client enables keepalive, the problem doesn't occur , e.g. `ab` with `-k`.

Maybe, Puma with SSL has a bug. I will check it later.

**Comment by jamesweb1** [comment]
Hi, thanks for your answering.
I found the argument with "-k" make the performance much better from 250 request/s to 3000 requests/s. However, I think it doesn't make sense in the real world for my need. In my practice, clients possibly come from different location and change with time, so they don't always need keepalive. Thanks your understanding again.

**Comment by repeatedly** [comment]
Open an issue on Puma: https://github.com/puma/puma/issues/567

**Comment by repeatedly** [comment]
Just released v0.0.2 with puma gem update
